design-proposals/feature-lifecycle.md

```diff
@@ -0,0 +1,327 @@
+and supported, something that does not exist with minor releases.
+
+#### The Deprecation Flow (for Minor releases)
+Only Alpha and Beta features can be removed during a minor release.
+
+These are the steps needed to deprecate & remove a feature:
+- Proposal: Prepare a proposal to remove a feature with proper
+  reasoning, phases, exact timelines and functional alternatives (if any).
+  The proposal should be reviewed and approved.
+- Notification: Notify the project community about the feature
+  discontinuation based on the approved proposal.
+  All details of the plan should be provided to allow users and possibly
+  down-stream projects to adjust.
+  Use all community media options to distribute this information
+  (e.g. mailing list, slack channel, community meetings).
+- Deprecation warnings: Add deprecation warnings at runtime to warn users
+  that the feature is planned to be removed.
+  Warnings should be raised when:
+  - Feature API fields are accessed.
+  - Feature FG is activated.
+  - Behavior related to the feature is detected (optional).
+- Removal: Feature removal involves removing the core functionality
+  of a feature and its exposed API.
+  - The core implementation can be removed in two steps:
+    - The FG is removed by assuring it is never reported as set
+      (i.e. even if it is left by the operator configured, internally
+       it is ignored).
+      At this stage, the core implementation will follow the FG conditions
+      and therefore from the outside the feature is inactive.
+    - In case there are no side effects, the core implementation code can
+      be removed.
+  - The API types are not to be removed, as it may have implications
+    with the underlying storage which has already persisted them.
+    Kubernetes has not removed fields, it just kept them around with the
+    warning that they have been deprecated and no longer available.
+
+    While keeping fields around for a period of a release or two makes
+    sense, beyond a limited period it adds a burden on dragging leftover
+    fields around to eternity.
+
+> **Note**: The only reference seen on why fields should not be removed
+> was mentioned [here](https://github.com/kubernetes/kubernetes/issues/52185).
+> But it is unclear if this is relevant for Alpha stage features.
+> Starting with a strict policy, similar to Kubernetes is recommended,
+> i.e. once fields are introduced, they should not be removed no matter
+> the feature release stage.
+> Per need, the topic can be revisited in follow-up adjustments.
+
+### Exceptions
+While the project strives to maintain a stable contract with its users,
+there may be scenarios where the policy described here will not be a fit.
+
+Therefore, it should be acceptable to have exceptions from time to time
+given a very good reasoning and an agreement from 2/3 of the project
+maintainers (also known as "approvers").
+
+## Implementation Phases
```

Comment:
Do we also need to align this effort with an actual release cycle as we are trying to do with the new and slightly overlapping enhancement process?

IOW saying that new features landing in 1.4 will need to adhere to this lifecycle?

Reply:
Let's revisit the timeline once we have agreement on the proposal.
After all to me 1.4 so far looks like a reasonable goal.